### ISSUE TITLE
503 Response being returned from Shinyproxy to Loadbalancer

### ISSUE THREAD

**Issue by mcrmonkey** [openanalytics/shinyproxy#65]
I have a shinyproxy sat behind a google load balancer and when accessing one of the configured applications the proxy will return a 503 response when some assets provided by the application are accessed.

The assets that fail aren't always the same though. Sometimes the logo on the applications start page doesn't load then it will when the page is reloaded and the browser cache ignored. The same can be said for CSS documents too.

The Loadbalancer is provided via the Google compute service and is configured to handle the SSL stuff for the proxy. Everything behind the loadbalancer for the moment is just HTTP

My shinyproxy config is:

```yaml

shiny:
  proxy:
    title: Test-shiny
    logo-url: https://assets.local/logo.png
    landing-page: /
    heartbeat-rate: 1000
    heartbeat-timeout: 1000000
    container-wait-time: 1200000
    hide-navbar: true
    port: 8080
    authentication: none
    docker:
      url: http://localhost:2375
      port-range-start: 20000
  apps:
  - name: test-app
    display-name: test-application
    description: test application
    docker-image: test-r-app:latest

logging:
  file:
    shinyproxy.log
  level:
    root: DEBUG

```

I have packet captures from inside the docker network that show all is well and the request is served properly.
The packet captures i have from shinyproxy however show its returning a 503 to the loadbalancer IP's

Network diagram of sorts:

{World} --> [Google LB] --> [shiny proxy instance  (Docker service) ]

If the LB is turned off shiny seems to be fine, however i kind of need it for what i want to do

Does there need to be extra config set for shiny to make it play nice with a loadbalancer ?

**Comment by fmichielssen** [comment]
Hi @mcrmonkey ,

Are you running multiple shinyproxy instances behind the load balancer?
And could you provide an example of a request URI that gets a 503 response?

**Comment by mcrmonkey** [comment]
Hi @fmichielssen 

Yes, the idea is to have multiple shinyproxy's but the project is/was going to rely on the loadbalancers ability to steer the user back to the correct place. 
right now for the purposes of testing there is 1 shinyproxy sat behind the LB and the 503's are still there.

An example of a URI would be: https://xxx.xxx.xxx.xxx/reverent_newton/thumbnail.png

The 503's generally only show up when loading the application via the browser. When accessing the assets directly they work every time even when hammering it a fair few times per second.

**Comment by mcrmonkey** [comment]
Thinking out loud here...

When using a google LB the `X-Forwarded-for` field contains two IP's ( comma separated ) 
the first being the IP address of the person accessing the app and the second being one of the many google LB IP's

Would this have any bearing on the access rights mechanism inside the web server component of the proxy? ( i.e. two IP's in the field messing with the logic )

**Comment by itssimon** [comment]
I'm having the exact same issue behind an AWS application load balancer. As in OPs case, the load balancer handles HTTPS and ShinyProxy serves everything via HTTP. Did anyone find a solution to this?

With AWS load balancers the `X-Forwarded-For` header only contains one IP - the client's.

**Comment by andrewsali** [comment]
We had a similar issue with nginx (openresty), setting a rate limit (20r/s) resolved the issue

**Comment by fmichielssen** [comment]
To all who are experiencing this issue: is your load balancer configured to support HTTP/2 ? If so, can you try disabling and testing if the issue persists?

**Comment by shrektan** [comment]
I think enabling  the sticky session in the load balancer may fix this...

**Comment by senfbrot** [comment]
@fmichielssen : disabling HTTP/2 at the load balancer indeed solves the 503 issue, the resources are properly returned then.

The load balancer itself does http1.1 requests towards the Shinyproxy in both cases, it seems as if HTTP/2 is active, then the requests from the load balancer towards Shinyproxy are "to fast" to be handled.

Maybe this is helpful, if HTTP/2 is active, then Shinyproxy reports the following error:
```
2019-04-16 14:08:52.303 ERROR 1 --- [   XNIO-2 I/O-2] org.xnio.nio                             : XNIO000011: Task io.undertow.server.handlers.proxy.ProxyHandler$ProxyClientHandler@60d73b7a failed with an exception

java.lang.IllegalStateException: UT000004: getResponseChannel() has already been called
	at io.undertow.server.HttpServerExchange.addResponseWrapper(HttpServerExchange.java:1463) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at io.undertow.server.HttpServerExchange.addResponseCommitListener(HttpServerExchange.java:1834) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at eu.openanalytics.containerproxy.util.ProxyMappingManager$1.getConnection(ProxyMappingManager.java:70) ~[containerproxy-0.6.0.jar!/:0.6.0]
	at io.undertow.server.handlers.proxy.ProxyHandler$ProxyClientHandler.failed(ProxyHandler.java:329) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at io.undertow.server.handlers.proxy.ProxyHandler$ProxyClientHandler.queuedRequestFailed(ProxyHandler.java:341) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at io.undertow.server.handlers.proxy.ProxyConnectionPool.connect(ProxyConnectionPool.java:531) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at io.undertow.server.handlers.proxy.LoadBalancingProxyClient.getConnection(LoadBalancingProxyClient.java:301) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at eu.openanalytics.containerproxy.util.ProxyMappingManager$1.getConnection(ProxyMappingManager.java:69) ~[containerproxy-0.6.0.jar!/:0.6.0]
	at io.undertow.server.handlers.proxy.ProxyHandler$ProxyClientHandler.run(ProxyHandler.java:310) ~[undertow-core-1.4.22.Final.jar!/:1.4.22.Final]
	at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:582) [xnio-nio-3.3.8.Final.jar!/:3.3.8.Final]
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:466) [xnio-nio-3.3.8.Final.jar!/:3.3.8.Final]
```


@shrektan : Sticky session had no effect

**Comment by mcrmonkey** [comment]
Can the speed of shineyproxy not be increased instead of dancing around it ?

**Comment by tverbeke** [comment]
@mcrmonkey It is the speed of the underlying Shiny technology that is problematic, not ShinyProxy itself. Note that starting from version 1.3.0 of Shiny (see [NEWS file](https://github.com/rstudio/shiny/blob/master/NEWS.md))

> Shiny now serves static files on a background thread. This means that things like JavaScript and CSS assets can be served without blocking or being blocked by the main R thread, and should result in significantly better performance for heavily loaded servers.

**Comment by senfbrot** [comment]
@tverbeke : same with Shiny 1.3.1 & HTTP/2

![image](https://user-images.githubusercontent.com/38048190/56288762-b7615300-611f-11e9-9b14-ca02e678c183.png)

**Comment by nysthee** [comment]
Hi any workarounds? We are running on GCP and we can not control HTTP version on the load balancer.
Our setup: 
- Shinyproxy in a container on K8S
- Behind a GCE ingress (default one)
- This ingress uses a GCP load balancer.

**Comment by ComeMaes** [comment]
Hi @nysthee,

I've got a similar setup, I'm disabling http2 on the ingress controller by linking it to a custom config map with the `use-http2: false` option, and it works fine.

**Comment by nysthee** [comment]
@ComeMaes Thanks for the pointer! Will try this out.

**Comment by k----n** [comment]
I can report that this is still an issue with the latest shiny.

Any find an issue on the shiny repo detailing the http/2 problem?

**Comment by johan13** [comment]
I am experiencing the same problem: Intermittent 503s behind AWS AppELB.

I have captured a failed request with tcpdump and ELB is already using http/1.1 when talking to ShinyProxy. (The request starts with: `GET /app_direct/ftg/shared/strftime/strftime-min.js HTTP/1.1`.)

I am the only user, so I doubt that is is a capacity issue. The Shiny app can do several thousand requests per second when I hammer it with `ab`, so setting a rate limit at 20/s as suggested above does not seem like a good solution.

I tried launching ShinyProxy with the environment variable `logging.requestdump=true` and the 503 errors are visible there as well. After a series of 200-responses I get a bunch of 503-responses before it goes back to returning 200s.

The log is below. (with lots of redactions)

Has anyone found a solution to this? Or is there anything else I can try to help pinpoint the problem?

```
==============================================================
2020-06-24 16:03:07.576  INFO 1 --- [   XNIO-2 I/O-1] io.undertow.request.dump                 : 
----------------------------REQUEST---------------------------
               URI=/app_direct/ftg/shinydashboard-0.7.1/shinydashboard.css
 characterEncoding=null
     contentLength=-1
       contentType=null
            cookie=JSESSIONID=...
            cookie=_fbp=...
            cookie=_ga=...
            cookie=_gid=...
            cookie=experimentation_subject_id=...
            header=cache-control=no-cache
            header=accept-encoding=gzip, deflate, br
            header=dnt=1
            header=pragma=no-cache
            header=sec-fetch-site=same-origin
            header=X-Forwarded-Port=443
            header=X-Forwarded-For=X.X.X.X
            header=cookie=experimentation_subject_id=... _ga=...; _fbp=...; _gid=...; JSESSIONID=d...
            header=referer=https://xxxx.com/app_direct/ftg/
            header=Host=xxxx.com
            header=accept=text/css,*/*;q=0.1
            header=X-Amzn-Trace-Id=Root=...
            header=accept-language=en,sv;q=0.9,de;q=0.8
            header=sec-fetch-mode=no-cors
            header=user-agent=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) snap Chromium/83.0.4103.106 Chrome/83.0.4103.106 Safari/537.36
            header=sec-fetch-dest=style
            header=X-Forwarded-Proto=https
            locale=[en, sv, de]
            method=GET
          protocol=HTTP/1.1
       queryString=
        remoteAddr=X.X.X.X:0
        remoteHost=X.X.X.X
            scheme=https
              host=xxxx.com
        serverPort=8080
--------------------------RESPONSE--------------------------
     contentLength=0
       contentType=null
            header=Expires=0
            header=Cache-Control=no-cache, no-store, max-age=0, must-revalidate
            header=X-XSS-Protection=1; mode=block
            header=Pragma=no-cache
            header=Date=Wed, 24 Jun 2020 16:03:07 GMT
            header=Connection=keep-alive
            header=X-Content-Type-Options=nosniff
            header=Strict-Transport-Security=max-age=31536000 ; includeSubDomains
            header=Content-Length=0
            status=503
```

**Comment by johan13** [comment]
I have been running some more `tcpdump` and it appears that requests that give 503-responses are not forwarded to Shiny at all - the 503s originate somewhere in ShinyProxy.

**Comment by LEDfan** [comment]
This should be fixed in ShinyProxy [2.4.0](https://shinyproxy.io/downloads/#240), therefore I'm closing this issue.
@johan13 @k----n @nysthee (and others) don't hesitate to re-open the issue (or open a new issue) if the problem persists.
